---
title: "Generate a 2-panel figure for R01 resubmission"
output: html_document
---

Detection of multi-strain colonization using lineage-defining variants curated from sequenced clones will be highly robust to sequencing errors and false-positive variant calls due to the strict criteria that known sets of variants at specific positions be present to confirm the presence of a strain.

Previously I have identified ~385 LDVs that differentiate nodes on the MI NH VREfm phylogeny.
Here we are testing additional scenarios:

To this end, can you please set up another set of simulations using three strains and the same metgenome sample (M):
6648-4103-0-HVRE (A)
5261-4089-0-RVRE (B)
1638-6008-0-RVRE (C)
Simulations:
A in M at 10%, 5%, 1%, 0.1%, 0.01%, 0.001% abundance
B in M at 10%, 5%, 1%, 0.1%, 0.01%, 0.001% abundance
C in M at 10%, 5%, 1%, 0.1%, 0.01%, 0.001% abundance
A + B in M at 10% abundance (5% each)
A + C in M at 10% abundance (5% each)
B + C in M at 10% abundance (5% each)
A + B + C in M at 10% abundance (3.3% each)

On June 25, Ali found some calculation errors in the previous spike-in experiment (using relative abundance in relatino to WGS depth instead of metagenome depth). He has since re-run the analysis and here is the updated result.

```{r load packages}
library(treeio)
library(ape)
library(ggtree)
library(ggplot2)
library(heatmap3)
library(randomcoloR)
library(phytools)
library(geiger)
library(cowplot)
library(RColorBrewer)
library(geiger)
library(vcfR)
library(plotfunctions)

knitr::opts_chunk$set(echo = FALSE)
```


```{r load data, include=FALSE}
dna = read.dna("../genome_alignments/2021_06_12_16_42_32_Aus0004_genome_aln_w_alt_allele_unmapped_exclude_controls.filtered_polymorphic_sites.fasta", format = "fasta")
var = read.table("../genome_alignments/2021_06_12_16_42_32_Aus0004_genome_aln_w_alt_allele_unmapped_exclude_controls.summary_of_snp_distribution.vcf")

dna2 = dna
colnames(dna2) = var$V2
dna2 = as.character(dna2)
dna2 = apply(dna2, 1:2, toupper)
var_pos2 = dna2[,apply(dna2, 2, FUN = function(x){sum(x %in% c("N", "-")) == 0})]

iqtree = read.iqtree("../genome_alignments/2021_06_12_16_42_32_Aus0004_genome_aln_w_alt_allele_unmapped_exclude_controls.treefile")

phylo = iqtree@phylo
phylo$bootstrap = phylo$node.label
phylo$node.label = c(min(phylo$edge[phylo$edge[,2] > length(phylo$tip.label),2]) - 1, phylo$edge[phylo$edge[,2] > length(phylo$tip.label),2])

# remove N and -
var_pos2 = var_pos2[,apply(var_pos2, 2, FUN = function(x){sum(x %in% c("N", "-")) == 0})]

# Load in data from different scenarios
rel_ab_list = list()
rel_ab_list[["A0.1"]] = read.csv("../genome_alignments/MI_simulations_v2/A_LDV_abund_frequency_0.1perc.csv", stringsAsFactors = F)
rel_ab_list[["A0.25"]] = read.csv("../genome_alignments/MI_simulations_v2/A_LDV_abund_frequency_0.25perc.csv", stringsAsFactors = F)
rel_ab_list[["A0.5"]] = read.csv("../genome_alignments/MI_simulations_v2/A_LDV_abund_frequency_0.5perc.csv", stringsAsFactors = F)
rel_ab_list[["A1"]] = read.csv("../genome_alignments/MI_simulations_v2/A_LDV_abund_frequency_1perc.csv", stringsAsFactors = F)
rel_ab_list[["A3"]] = read.csv("../genome_alignments/MI_simulations_v2/A_LDV_abund_frequency_3perc.csv", stringsAsFactors = F)
rel_ab_list[["A5"]] = read.csv("../genome_alignments/MI_simulations_v2/A_LDV_abund_frequency_5perc.csv", stringsAsFactors = F)
rel_ab_list[["A10"]] = read.csv("../genome_alignments/MI_simulations_v2/A_LDV_abund_frequency_10perc.csv", stringsAsFactors = F)

rel_ab_list[["B0.1"]] = read.csv("../genome_alignments/MI_simulations_v2/B_LDV_abund_frequency_0.1perc.csv", stringsAsFactors = F)
rel_ab_list[["B0.25"]] = read.csv("../genome_alignments/MI_simulations_v2/B_LDV_abund_frequency_0.25perc.csv", stringsAsFactors = F)
rel_ab_list[["B0.5"]] = read.csv("../genome_alignments/MI_simulations_v2/B_LDV_abund_frequency_0.5perc.csv", stringsAsFactors = F)
rel_ab_list[["B1"]] = read.csv("../genome_alignments/MI_simulations_v2/B_LDV_abund_frequency_1perc.csv", stringsAsFactors = F)
rel_ab_list[["B3"]] = read.csv("../genome_alignments/MI_simulations_v2/B_LDV_abund_frequency_3perc.csv", stringsAsFactors = F)
rel_ab_list[["B5"]] = read.csv("../genome_alignments/MI_simulations_v2/B_LDV_abund_frequency_5perc.csv", stringsAsFactors = F)
rel_ab_list[["B10"]] = read.csv("../genome_alignments/MI_simulations_v2/B_LDV_abund_frequency_10perc.csv", stringsAsFactors = F)

rel_ab_list[["C0.1"]] = read.csv("../genome_alignments/MI_simulations_v2/C_LDV_abund_frequency_0.1perc.csv", stringsAsFactors = F)
rel_ab_list[["C0.25"]] = read.csv("../genome_alignments/MI_simulations_v2/C_LDV_abund_frequency_0.25perc.csv", stringsAsFactors = F)
rel_ab_list[["C0.5"]] = read.csv("../genome_alignments/MI_simulations_v2/C_LDV_abund_frequency_0.5perc.csv", stringsAsFactors = F)
rel_ab_list[["C1"]] = read.csv("../genome_alignments/MI_simulations_v2/C_LDV_abund_frequency_1perc.csv", stringsAsFactors = F)
rel_ab_list[["C3"]] = read.csv("../genome_alignments/MI_simulations_v2/C_LDV_abund_frequency_3perc.csv", stringsAsFactors = F)
rel_ab_list[["C5"]] = read.csv("../genome_alignments/MI_simulations_v2/C_LDV_abund_frequency_5perc.csv", stringsAsFactors = F)
rel_ab_list[["C10"]] = read.csv("../genome_alignments/MI_simulations_v2/C_LDV_abund_frequency_10perc.csv", stringsAsFactors = F)

rel_ab_list[["AB"]] = read.csv("../genome_alignments/MI_simulations_v2/A_B_in_M_LDV_abund_frequency.csv", stringsAsFactors = F)
rel_ab_list[["AC"]] = read.csv("../genome_alignments/MI_simulations_v2/A_C_in_M_LDV_abund_frequency.csv", stringsAsFactors = F)
rel_ab_list[["BC"]] = read.csv("../genome_alignments/MI_simulations_v2/B_C_in_M_LDV_abund_frequency.csv", stringsAsFactors = F)
rel_ab_list[["ABC"]] = read.csv("../genome_alignments/MI_simulations_v2/A_B_C_in_M_LDV_abund_frequency.csv", stringsAsFactors = F)

```

Determine LDVs based on phylogeny
```{r generate subtrees}
vrefm_st = subtrees(phylo)

vrefm_st_vars = lapply(vrefm_st, FUN = function(l){
  
  temp_mat = var_pos2[l$tip.label,]

  shared = apply(temp_mat, 2 , FUN = function(x){length(unique(x)) == 1})
  
  uniq_snv = apply(temp_mat[,names(which(shared))] , 2, FUN = function(x){unique(x)})
  
  uniq_snv
})

vrefm_st_ind = 1:length(vrefm_st) 

# specify nucleotide sequence identity

vrefm_st_ldv_id = list()

for (s in vrefm_st_ind){print(s)
  
  test_samples = vrefm_st[[s]]$tip.label
  
  test_st = vrefm_st_vars[[s]]
  
  test_st_uniq_var = sapply(names(test_st), FUN = function(r){
    
  test_snv = var_pos2[test_samples, r]
  
  if (unique(test_snv) != test_st[r]){stop} # extra check
  
  sum(unique(test_snv) %in% names(table(var_pos2[!rownames(var_pos2) %in% test_samples,r]))) == 0
    
  })
  
  if (length(test_st_uniq_var) == 0){vrefm_st_ldv_id[[as.character(s)]] = NULL}else{
  vrefm_st_ldv_id[[as.character(s)]] = structure(test_st[names(which(test_st_uniq_var))], names = names(which(test_st_uniq_var)))}
}

nodes_with_LDVs = names(which(lapply(vrefm_st_ldv_id, length) > 0))
nodes_without_LDVs = vrefm_st_ind[!vrefm_st_ind %in% nodes_with_LDVs]

vrefm_st_ldv = sapply(names(unlist(vrefm_st_ldv_id)), FUN = function(x){strsplit(x = x, split = ".", fixed = T)[[1]][2]})
```

# Quantify LDV-detecting sensitivity at various relative abundances
Also for panel B, what does it look like if instead of %LDVs, you plot the percent of expected nodes with >50% of LDVs detected?
```{r}
i = c("6648-4103-0-HVRE", "5261-4089-0-RVRE", "1638-6008-0-RVRE")
i_list = list(A = c("A10", "A5", "A3", "A1", "A0.5", "A0.25", "A0.1"),
              B = c("B10", "B5", "B3", "B1", "B0.5", "B0.25", "B0.1"),
              C = c("C10", "C5", "C3", "C1", "C0.5", "C0.25", "C0.1"))

sample_list = list()
for (p in i){
  
  if (p == i[1]){samples = i_list$A}
  if (p == i[2]){samples = i_list$B}
  if (p == i[3]){samples = i_list$C}
  
  pct_expected_nodes_with_gt50_ldvs_detected = list()

  for (q in samples){
  
  i_present_st = which(unlist(lapply(vrefm_st, FUN = function(x){sum(grepl(p, x$tip.label)) > 0})))
  ldv_i_present_st_ind = intersect(i_present_st, nodes_with_LDVs)
  i_specific_ldvs = structure(unlist(vrefm_st_ldv_id[ldv_i_present_st_ind]), names =
                              sapply(names(unlist(vrefm_st_ldv_id[ldv_i_present_st_ind])), FUN = function(x){
                                strsplit(x, ".", fixed = T)[[1]][2]
                              }))

  temp_df = rel_ab_list[[q]]
  
  colnames(temp_df) = c("LDV Position", "Strain Position", "Ref Allele", "Ref Allele Freq", "ALT Allele", "ALT Allele Freq", "ALT Allele Depth")
  
  temp_df = temp_df[nchar(temp_df$`Ref Allele`) == 1 & nchar(temp_df$`ALT Allele`) == 1,]
  
  # subset to only LDV-associated positions
  ldv_temp_df = temp_df[temp_df$`LDV Position` %in% names(i_specific_ldvs), ]
  
  # reorganize df to merge duplicate positions
  
  temp_df_pos = sapply(unique(ldv_temp_df$`LDV Position`), FUN = function(l){
    
    alleles = c(structure(ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"Ref Allele Freq"], 
                        names = ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"Ref Allele"]), 
                  structure(ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"ALT Allele Freq"], 
                            names = ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"ALT Allele"]))
    
    allele_present = paste(names(which(alleles > 0)), collapse = "/")
    allele_present
  })
  names(temp_df_pos) = unique(ldv_temp_df$`LDV Position`)

  # count number of nodes with any LDV-supporting alleles
  test_node_gt50 = sapply(ldv_i_present_st_ind, FUN = function(d){
    st_ldv = vrefm_st_ldv_id[[d]]
    st_ldv_pos = names(st_ldv)
    sum(temp_df_pos[st_ldv_pos] == st_ldv, na.rm = T) / length(st_ldv) >= 0.5
  })

 pct_expected_nodes_with_gt50_ldvs_detected[[q]] = test_node_gt50 

}
sample_list[[p]] = pct_expected_nodes_with_gt50_ldvs_detected
}
```


# plot as scatterplots
```{r}
file = paste("../figures/", format(Sys.time(), "%Y-%m-%d"), '_R01_panel_figures_new_spike_conc.pdf', sep = "");
pdf(file, width = 14)
par(mfrow = c(1,2))
# Tree with mixed isolates

  temp_df = rel_ab_list$ABC
  
  colnames(temp_df) = c("LDV Position", "Strain Position", "Ref Allele", "Ref Allele Freq", "ALT Allele", "ALT Allele Freq")
  
  temp_df = temp_df[nchar(temp_df$`Ref Allele`) == 1 & nchar(temp_df$`ALT Allele`) == 1,]
  
  # subset to only LDV-associated positions
  ldv_temp_df = temp_df[temp_df$`LDV Position` %in% vrefm_st_ldv, ]
  
  # reorganize df to merge duplicate positions
  
  temp_df_pos = sapply(unique(ldv_temp_df$`LDV Position`), FUN = function(l){
    
  alleles = c(structure(ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"Ref Allele Freq"], names = ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"Ref Allele"]), 
                  structure(ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"ALT Allele Freq"], names = ldv_temp_df[ldv_temp_df$`LDV Position` %in% l,"ALT Allele"]))
    
    allele_present = paste(names(which(alleles > 0)), collapse = "/")
    allele_present
  })
  names(temp_df_pos) = unique(ldv_temp_df$`LDV Position`)
  
test_node = lapply(vrefm_st_ldv_id, FUN = function(v){print(v)
  test_pos = sapply(names(v), FUN = function(p){print(p)
   grepl(v[p], temp_df_pos[p])
  })
  sum(unlist(test_pos), na.rm = T) / length(unlist(test_pos))
})


# Then repeat the tree exercise to see where the nodes are
test_group_clade = names(which(unlist(lapply(test_node, FUN = function(n){n > 0}))))
test_phylo_group_clade = unlist(lapply(vrefm_st[as.numeric(test_group_clade)], FUN = function(x){x$name}))

test_node_ind = unlist(test_node)
test_node_ind = test_node_ind[!is.na(test_node_ind) & test_node_ind > 0]

nodecol_ind = structure(colorRampPalette(c("grey", "red"))(101), names = 0:100)
nodecol = rep("#808080", phylo$Nnode)
nodecol[as.numeric(names(test_node_ind))] = unlist(sapply(test_node_ind, FUN = function(d){nodecol_ind[as.character(round(d*100))]}))
nodecol[nodes_without_LDVs] = "black"

tipcol = rep("transparent", length(phylo$tip.label))
tipcol[phylo$tip.label %in% "6648-4103-0-HVRE"] = "purple"
tipcol[phylo$tip.label %in% "5261-4089-0-RVRE"] = "blue"
tipcol[phylo$tip.label %in% "1638-6008-0-RVRE"] = "orange"

plot(phylo, use.edge.length = T,edge.color = "black",  cex.main = 2, show.tip.label =F, pch =16, main = "Mixed colonization (in silico simulation)")
tiplabels(pch = 21, col = tipcol, cex = 1.5, bg = tipcol)
nodelabels(pch = 21, col = "transparent", cex = .75, bg = nodecol)
gradientLegend(valRange = c(0, 100), pos = 0.5, side = 1, color = nodecol_ind, n.seg = 9, border.col = NA, dec = 0, length = .75)
legend(x = "topright", legend = paste0("Strain ", 1:3), col = c("purple", "blue", "orange"), fill = c("purple", "blue", "orange"), bty = "n", cex = 2)

# Scatterplot showing lineage-specific LDVs

pch_shape = structure(c(15, 17, 19), names = names(sample_list))
pch_col = structure(c("purple", "blue", "orange"), names = names(sample_list))

plot(0, 0, xlab = "Spike-in relative abundance (%)", ylab = "Proportion", xaxt = "n", yaxt = "n", col = "white", xlim = c(1, 7), ylim = c(0, 1), main = "Proportion of nodes\nsupported by 50% LDVs", cex.main = 2, cex.lab = 1.5)
axis(side = 1, at = 1:7, labels = c("10", "5", "3", "1", "0.5", "0.25", "0.1"), cex.axis = 1)
axis(side = 2, at = seq(0, 1, by = .1), labels = seq(0, 1, by = .1), las= 2, cex.axis = 1)
for (p in names(sample_list)){
  lineage_ldv = unlist(lapply(sample_list[[p]], FUN = function(x){sum(x)/ length(x)}))
  points(lineage_ldv, type = "b", pch = pch_shape[p], col = pch_col[p], cex = 2)
}
legend(x = "bottomleft", legend = paste0("Strain ", 1:3), col = pch_col, pch = pch_shape, bty = "n", cex = 2)


dev.off()
```


