---
title: "Finding lineage-defining variants in Michigan nursing home VREfm"
output: html_document
---

Detection of multi-strain colonization using lineage-defining variants curated from sequenced clones will be highly robust to sequencing errors and false-positive variant calls due to the strict criteria that known sets of variants at specific positions be present to confirm the presence of a strain.

Ali has generated gubbins-filtered alignment and tree in:
/nfs/esnitkin/Project_VRE_metagenomics_analysis/Sequence_data/output_files/2021_04_09_VREfm_variant_calling/2021_04_12_15_54_54_core_results/gubbins/

Next, Ali spiked a metagenome sample from Krishna with a VREfm strain (_____), and mapped to Aus0004 using freebayes.
Now we want to see how the freebayes result compares with the "normal" WGS pipeline.
Also assess
1. Whether the 5261 allele is detected
2. What fraction of the reads support it

```{r load packages}
library(treeio)
library(ape)
library(ggtree)
library(ggplot2)
library(heatmap3)
library(randomcoloR)
library(phytools)
library(geiger)
library(cowplot)
library(RColorBrewer)
library(geiger)
library(vcfR)

knitr::opts_chunk$set(echo = FALSE)
```


```{r load data, include=FALSE}
vrefm = read.dna("../genome_alignments/2021_04_12_15_54_54_Aus0004_genome_aln_w_alt_allele_unmapped.filtered_polymorphic_sites.fasta", format = "fasta")

var_pos = read.csv("../genome_alignments/2021_04_12_15_54_54_Aus0004_genome_aln_w_alt_allele_unmapped.summary_of_snp_distribution.csv", stringsAsFactors = F, header = T, skip = 3)
var_pos = t(var_pos)
colnames(var_pos) = gsub(" ", "", as.character(var_pos["POS", ]))

var_pos2 = var_pos[grep("CHROM|POS|ID|REF|ALT|QUAL|FILTER|INFO|FORMAT", x = rownames(var_pos), invert = T),]
rownames(var_pos2) = gsub("X", "", rownames(var_pos2))
rownames(var_pos2) = gsub(".", "-", rownames(var_pos2), fixed = T)
rownames(var_pos2)[rownames(var_pos2) %in% "CP003351-1"] = "CP003351.1"

iqtree = read.iqtree("../genome_alignments/2021_04_12_15_54_54_Aus0004_genome_aln_w_alt_allele_unmapped.treefile")

phylo = iqtree@phylo
phylo$bootstrap = phylo$node.label
phylo$node.label = c(min(phylo$edge[phylo$edge[,2] > length(phylo$tip.label),2]) - 1, phylo$edge[phylo$edge[,2] > length(phylo$tip.label),2])

# remove N and -
var_pos2 = var_pos2[,apply(var_pos2, 2, FUN = function(x){sum(x %in% c("N", "-")) == 0})]

# total SNVs with FreeBayes
# pos = first column, Ref = fourth column, ALT = fifth column

total = read.vcfR("../genome_alignments/5261-4089-0-RVRE_Aus0004_with_metagenome_raw.vcf.gz")
rel_ab = read.csv("../genome_alignments/LDV_abund_frequency.csv", stringsAsFactors = F)
```

Check to see if all three files (gubbin filtered SNVs, VCF file, and LDV file) have the same number of sites
```{r}
var_pos2_ldv_diff = setdiff(colnames(var_pos2), rel_ab$LDV.Position)

var_pos2_ldv_same = intersect(colnames(var_pos2), rel_ab$LDV.Position)

rel_ab_df = rel_ab
rel_ab_df$alt_allele = apply(rel_ab, 1, FUN = function(r){
  ifelse(r["ALT.Allele"] %in% ".", r["Reference.Allele"], r["ALT.Allele"])
})

# two positions are duplicated - manually remove entries with lower freq for now
rel_ab_df = rel_ab_df[!rownames(rel_ab_df) %in% c(434986, 1257162), ]

var_pos2_ldv_check = rel_ab_df[rel_ab_df$LDV.Position %in% as.character(var_pos2_ldv_same), "alt_allele"] == var_pos2["5261-4089-0-RVRE",as.character(var_pos2_ldv_same)]

var_pos2_ldv_check_ref_alt = cbind(rel_ab_df[rel_ab_df$LDV.Position %in% names(which(!var_pos2_ldv_check)), c("LDV.Position","Reference.Allele", "Reference.Allele.Frequency", "ALT.Allele", "ALT.Allele.Frequency")], var_pos2["5261-4089-0-RVRE", names(which(!var_pos2_ldv_check))])

paste0("VCF file contains ", length(unique(total@fix[,"POS"])), " unique positions")
paste0("LDV abundance file has ", length(unique(rel_ab$LDV.Position)), " unique positions")
paste0(length(var_pos2_ldv_diff), " positions are present in WGS alignment but not LDV file")
paste0("Check LDV abundance sites against WGS alignment result (", ncol(var_pos2), " positions)")
paste0(round(sum(var_pos2_ldv_check) / length(var_pos2_ldv_check)*100, 2), "% of freebayes calls agree with WGS alignment")
paste0(sum(var_pos2_ldv_check_ref_alt$Reference.Allele == var_pos2_ldv_check_ref_alt$`var_pos2["5261-4089-0-RVRE", names(which(!var_pos2_ldv_check))]`), " out of ", length(var_pos2_ldv_check_ref_alt$Reference.Allele == var_pos2_ldv_check_ref_alt$`var_pos2["5261-4089-0-RVRE", names(which(!var_pos2_ldv_check))]`),  " discrepant sites are identical to reference allele at an average frequence of ", round(mean(var_pos2_ldv_check_ref_alt[var_pos2_ldv_check_ref_alt$Reference.Allele == var_pos2_ldv_check_ref_alt$`var_pos2["5261-4089-0-RVRE", names(which(!var_pos2_ldv_check))]`,"Reference.Allele.Frequency"])*100), "%")
```


```{r generate subtrees}
vrefm_st = subtrees(phylo)

vrefm_st_vars = lapply(vrefm_st, FUN = function(l){
  
  temp_mat = var_pos2[l$tip.label,]

  shared = apply(temp_mat, 2 , FUN = function(x){length(unique(x)) == 1})
  
  uniq_snv = apply(temp_mat[,names(which(shared))] , 2, FUN = function(x){unique(x)})
  
  uniq_snv
})

vrefm_st_ind = 1:length(vrefm_st) # skip first subtree because there are no lineage-defining variants for the overall tree

vrefm_st_ldv = list()

for (s in vrefm_st_ind){print(s)
  
  test_samples = vrefm_st[[s]]$tip.label
  
  test_st = vrefm_st_vars[[s]]
  
  test_st_uniq_var = sapply(names(test_st), FUN = function(r){
    
  test_snv = var_pos2[test_samples, r]
  
  if (unique(test_snv) != test_st[r]){stop} # extra check
  
  sum(unique(test_snv) %in% names(table(var_pos2[!rownames(var_pos2) %in% test_samples,r]))) == 0
    
  })
  
  if (length(test_st_uniq_var) == 0){vrefm_st_ldv[[as.character(s)]] = NULL}else{
  vrefm_st_ldv[[as.character(s)]] = names(which(test_st_uniq_var))}
}


# isolate specific LDVs if ever needed:
i = "5261-4089-0-RVRE"

test_i = sapply(vrefm_st_ind, FUN = function(l){
  i_present = i %in% vrefm_st[[l]]$tip.label 
})
  
ldv_i = vrefm_st_ldv[as.character(which(test_i))]
```

Check to see if these LDVs are specific to this isolate


```{r}
all_ldvs = unlist(vrefm_st_ldv)

rel_ab_ldv = rel_ab_df[rel_ab_df$LDV.Position %in% all_ldvs,]


# file = paste("../figures/", format(Sys.time(), "%Y-%m-%d"), '_5261-Aus0004-LDV_alt_allele_freq.pdf', sep = "");
# pdf(file)

boxplot(rel_ab_ldv$ALT.Allele.Frequency*100 ~ (rel_ab_ldv$LDV.Position %in% unlist(ldv_i) == 1),
  xaxt = "n",
  xlab = "",
  beside = T,
  ylab = "Alternate allele frequency (%)", outline = F)

axis(side = 1, at = 1:2, 
     labels = c(paste0("non-LDVs (N = ", length(rel_ab_ldv[!rel_ab_ldv$LDV.Position %in% unlist(ldv_i), "ALT.Allele.Frequency"]), ")"), 
                paste0("LDVs (N = ", length(unlist(ldv_i)), ")")))

stripchart(
  rel_ab_ldv$ALT.Allele.Frequency*100 ~ (rel_ab_ldv$LDV.Position %in% unlist(ldv_i) == 1),
  group.names = c(paste0("LDVs (N = ", length(unlist(ldv_i)), ")"), 
                paste0("non-LDVs (N = ", length(rel_ab_ldv[!rel_ab_ldv$LDV.Position %in% unlist(ldv_i), "ALT.Allele.Frequency"]), ")")),
  vertical = T,
  ylab = "Alternate allele frequency (%)", add = T, col = 3:4, pch = 19)
# dev.off()
```

Plot tree to see which nodes are used for generating LDVs
```{r}
cols = colorRampPalette(c("white", "red"))(100)

group_clade = names(which(unlist(lapply(ldv_i, length)) > 0))

# drop the largest subtree
# group_clade2 = group_clade[!group_clade %in% as.character(min(as.numeric(group_clade)))]
group_clade2 = group_clade
phylo_group_clade = unlist(lapply(vrefm_st[as.numeric(group_clade2)], FUN = function(x){x$name}))
edgecol <- rep('black', nrow(phylo$edge))

phylo_group_clade_end = c(sapply(phylo_group_clade[1:(length(phylo_group_clade) -1)], FUN = function(s){
  s_end = ifelse(sum(phylo$edge[phylo$edge[,1] %in% s, 2] %in% phylo_group_clade), phylo$edge[phylo$edge[,1] %in% s, 2][which(phylo$edge[phylo$edge[,1] %in% s, 2] %in% phylo_group_clade)], s+1)
}), phylo_group_clade[length(phylo_group_clade)])
  
# phylo_group_clade_end = sapply(phylo_group_clade, FUN = function(s){
#   s_end = ifelse(sum(phylo$edge[phylo$edge[,1] %in% s, 2] %in% (which(phylo$tip.label %in% i))) >0 | (s+1) %in% phylo_group_clade | sum(tips(phylo, s+1) %in% i) == 0, s, s+1)
#   s_end
# })

edgecol_ind = apply(phylo$edge, 1, FUN = function(x){
  sum(apply(cbind(phylo_group_clade, phylo_group_clade_end), 1, FUN = function(y){
    x[1] == y[1] & x[2] == y[2]
  }))
})
edgecol[edgecol_ind] = "red"

tipcol = rep("black", length(phylo$tip.label))
tipcol[phylo$tip.label %in% i] = "brown"

# file = paste("../figures/", format(Sys.time(), "%Y-%m-%d"), '_5261-Aus0004-varcall-with-mg-with_node.pdf', sep = "");
# pdf(file)
plot(phylo, use.edge.length = T,edge.color = edgecol, tip.color = tipcol, cex = 0.25)
# dev.off()
```



If I knew nothing about this spiked isolate, how would I go about to finding its closest node?
```{r}
test_rel_ab_ldv = rel_ab_ldv[nchar(rel_ab_ldv$alt_allele) == 1, ] 

# specify nucleotide sequence identity

vrefm_st_ldv_id = list()

for (s in vrefm_st_ind){print(s)
  
  test_samples = vrefm_st[[s]]$tip.label
  
  test_st = vrefm_st_vars[[s]]
  
  test_st_uniq_var = sapply(names(test_st), FUN = function(r){
    
  test_snv = var_pos2[test_samples, r]
  
  if (unique(test_snv) != test_st[r]){stop} # extra check
  
  sum(unique(test_snv) %in% names(table(var_pos2[!rownames(var_pos2) %in% test_samples,r]))) == 0
    
  })
  
  if (length(test_st_uniq_var) == 0){vrefm_st_ldv_id[[as.character(s)]] = NULL}else{
  vrefm_st_ldv_id[[as.character(s)]] = structure(test_st[names(which(test_st_uniq_var))], names = names(which(test_st_uniq_var)))}
}

# vrefm_st_ldv_id = vrefm_st_ldv_id[unlist(lapply(vrefm_st_ldv_id, FUN = function(l){length(l) > 0}))]

test_node = lapply(vrefm_st_ldv_id, FUN = function(v){print(v)
  test_pos = sapply(names(v), FUN = function(p){print(p)
    test_rel_ab_ldv[test_rel_ab_ldv$LDV.Position %in% p, "alt_allele"] == v[p]
  })
  sum(unlist(test_pos)) / length(unlist(test_pos))
})

# Then repeat the tree exercise to see where the nodes are
# cols = structure(c("white", "red"), names = c("No", "Yes"))
test_group_clade = names(which(unlist(lapply(test_node, FUN = function(n){n > 0}))))
test_phylo_group_clade = unlist(lapply(vrefm_st[as.numeric(test_group_clade)], FUN = function(x){x$name}))

test_node_ind = unlist(test_node)
test_node_ind = test_node_ind[!is.na(test_node_ind) & test_node_ind > 0]

nodecol_ind = structure(colorRampPalette(c("grey", "red"))(length(c(0,unique(test_node_ind)))), names = as.character(c(0,sort(unique(test_node_ind)))))
nodecol = rep("#808080", phylo$Nnode)
nodecol[as.numeric(names(test_node_ind))] = sapply(test_node_ind, FUN = function(d){nodecol_ind[as.character(d)]})

tipcol = rep("black", length(phylo$tip.label))
tipcol[phylo$tip.label %in% i] = "orange"

file = paste("../figures/", format(Sys.time(), "%Y-%m-%d"), '_5261-Aus0004-varcall-with-mg-with_node.pdf', sep = "");
pdf(file)
plot(phylo, use.edge.length = T,edge.color = "black",  cex = 0.25, show.tip.label =F, pch =16)
tiplabels(pch = 21, col = tipcol, cex = .5, bg = tipcol)
nodelabels(pch = 21, col = "transparent", cex = 0.75, bg = nodecol)
legend("right", legend  = as.numeric(names(nodecol_ind))*100,
         fill = nodecol_ind, title = "% LDVs\ndetected", border = FALSE, bty = "n", cex = 1, inset =0, xpd = TRUE)

plot(phylo, use.edge.length = F,edge.color = "black",  cex = 0.25, show.tip.label =F, pch =16)
tiplabels(pch = 21, col = tipcol, cex = .5, bg = tipcol)
nodelabels(pch = 21, col = "transparent", cex = 0.75, bg = nodecol)
legend("bottom", legend  = as.numeric(names(nodecol_ind))*100,
         fill = nodecol_ind, title = "% LDVs detected", border = FALSE, bty = "n", cex = 1, 
       inset = -.1, xpd = TRUE, horiz = T)

dev.off()
``` 