---
title: "Finding lineage-defining variants in Michigan nursing home VREfm"
output: html_document
---

Detection of multi-strain colonization using lineage-defining variants curated from sequenced clones will be highly robust to sequencing errors and false-positive variant calls due to the strict criteria that known sets of variants at specific positions be present to confirm the presence of a strain.

Ali has generated gubbins-filtered alignment and tree in:
/nfs/esnitkin/Project_VRE_metagenomics_analysis/Sequence_data/output_files/2021_04_09_VREfm_variant_calling/2021_04_12_15_54_54_core_results/gubbins/

There are promising LDVs in each subtree, and after discussing with Evan, it seems that it's better to use a subtree-centric approach to assign isolates, rather than the reverse.

```{r load packages}
library(treeio)
library(ape)
library(ggtree)
library(ggplot2)
library(heatmap3)
library(randomcoloR)
library(phytools)
library(geiger)
library(cowplot)
library(RColorBrewer)
library(geiger)

knitr::opts_chunk$set(echo = FALSE)
```


```{r load data, include=FALSE}
vrefm = read.dna("../genome_alignments/2021_04_12_15_54_54_Aus0004_genome_aln_w_alt_allele_unmapped.filtered_polymorphic_sites.fasta", format = "fasta")

var_pos = read.csv("../genome_alignments/2021_04_12_15_54_54_Aus0004_genome_aln_w_alt_allele_unmapped.summary_of_snp_distribution.csv", stringsAsFactors = F, header = T, skip = 3)
var_pos = t(var_pos)
colnames(var_pos) = gsub(" ", "", as.character(var_pos["POS", ]))

var_pos2 = var_pos[grep("CHROM|POS|ID|REF|ALT|QUAL|FILTER|INFO|FORMAT", x = rownames(var_pos), invert = T),]
rownames(var_pos2) = gsub("X", "", rownames(var_pos2))
rownames(var_pos2) = gsub(".", "-", rownames(var_pos2), fixed = T)
rownames(var_pos2)[rownames(var_pos2) %in% "CP003351-1"] = "CP003351.1"

iqtree = read.iqtree("../genome_alignments/2021_04_12_15_54_54_Aus0004_genome_aln_w_alt_allele_unmapped.treefile")

phylo = iqtree@phylo
phylo$node.label = c(min(phylo$edge[phylo$edge[,2] > length(phylo$tip.label),2]) - 1, phylo$edge[phylo$edge[,2] > length(phylo$tip.label),2])

# remove N and -
var_pos2 = var_pos2[,apply(var_pos2, 2, FUN = function(x){sum(x %in% c("N", "-")) == 0})]
```

```{r generate subtrees}
vrefm_st = subtrees(phylo)

vrefm_st_vars = lapply(vrefm_st, FUN = function(l){
  
  temp_mat = var_pos2[l$tip.label,]

  shared = apply(temp_mat, 2 , FUN = function(x){length(unique(x)) == 1})
  
  uniq_snv = apply(temp_mat[,names(which(shared))] , 2, FUN = function(x){unique(x)})
  
  uniq_snv
})

vrefm_st_ind = 1:length(vrefm_st) # skip first subtree because there are no lineage-defining variants for the overall tree

vrefm_st_ldv = list()

for (s in vrefm_st_ind){print(s)
  
  test_samples = vrefm_st[[s]]$tip.label
  
  test_st = vrefm_st_vars[[s]]
  
  test_st_uniq_var = sapply(names(test_st), FUN = function(r){
    
  test_snv = var_pos2[test_samples, r]
  
  if (unique(test_snv) != test_st[r]){stop} # extra check
  
  sum(unique(test_snv) %in% names(table(var_pos2[!rownames(var_pos2) %in% test_samples,r]))) == 0
    
  })
  
  if (length(test_st_uniq_var) == 0){vrefm_st_ldv[[as.character(s)]] = NULL}else{
  vrefm_st_ldv[[as.character(s)]] = names(which(test_st_uniq_var))}
}


# isolate specific LDVs if ever needed:
i = "5261-4089-0-RVRE"

test_i = sapply(vrefm_st_ind, FUN = function(l){
  i_present = i %in% vrefm_st[[l]]$tip.label
})
  
ldv_i = vrefm_st_ldv[as.character(which(test_i)[!which(test_i) %in% 1])]
```

Ali finished his pipeline; output file in
/nfs/esnitkin/Project_VRE_metagenomics_analysis/Analysis/2021_05_07_LDV_test/5261-4089-0-RVRE_same_assembly_varcall_result/LDV_abund_frequency.csv

```{r}
output = read.csv("../genome_alignments/spikein_expt/5261-4089-0-RVRE_Aus0004_with_metagenome_varcall_result/LDV_abund_frequency.csv", header = T)

output_df = as.data.frame(output)
output_df = output_df[output_df$LDV.Position %in% unlist(vrefm_st_ldv), ]
ldv_i_pos = unlist(ldv_i)
output_df[output_df$LDV.Position %in% ldv_i_pos,]

# isolate alleles 100% frequency
i_ldv_100 = nrow(output_df[output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency %in% 1,]) / length(ldv_i_pos)

# isolate alleles less than 100% frequency
i_ldv_lt100 = nrow(output_df[output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency < 1 &output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency  > 0 ,]) / length(ldv_i_pos)

# non-isolate alleles 100% frequency
non_i_ldv_100 = nrow(output_df[!output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency %in% 1,]) / nrow(output_df[!output_df$LDV.Position %in% ldv_i_pos, ])

# non-isolate alleles less than 100% frequency
non_i_ldv_lt100 = nrow(output_df[!output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency < 1 & !output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency > 0,]) / nrow(output_df[!output_df$LDV.Position %in% ldv_i_pos, ])

# Check to see if these LDVs are specific to this isolate
group_clade= names(which(unlist(lapply(ldv_i, length)) > 0))

phylo_group_clade = as.numeric(group_clade) + length(phylo$tip.label) # need to add one because node is number of tips + 1
edgecol <- rep('black', nrow(phylo$edge))

phylo_group_clade_end = sapply(phylo_group_clade, FUN = function(s){
  s_end = ifelse(sum(phylo$edge[phylo$edge[,1] %in% s, 2] %in% (which(phylo$tip.label %in% i))) >0 | (s+1) %in% phylo_group_clade | sum(tips(phylo, s+1) %in% i) == 0, s, s+1)
  s_end
})

edgecol[phylo$edge[,1] %in% phylo_group_clade & (phylo$edge[,2] %in% union(which(phylo$tip.label %in% i), c(phylo_group_clade, phylo_group_clade_end)))] = "red"
tipcol = rep("black", length(phylo$tip.label))
tipcol[phylo$tip.label %in% i] = "brown"

# Proportion of supporting LDV by node
ldv_i_mat = matrix(0, nrow = 3, ncol = sum(unlist(lapply(ldv_i, length)) > 0), dimnames = list(c("Total LDV", "Alt Allele at 100%", "%"), 1:sum(unlist(lapply(ldv_i, length)) > 0)))

ldv_i_mat[1,] = unlist(lapply(ldv_i[names(which(unlist(lapply(ldv_i, length)) > 0))], length))
ldv_i_mat[2,] = unlist(lapply(ldv_i[names(which(unlist(lapply(ldv_i, length)) > 0))], FUN = function(l){
    sum(output_df[output_df$LDV.Position %in% l,"ALT.Allele.Frequency"] == 1)
}))
ldv_i_mat[3,] = (ldv_i_mat[2,] / ldv_i_mat[1,]) * 100

```

```{r plot}
# with node labels
file = paste("../figures/", format(Sys.time(), "%Y-%m-%d"), '_5261-Aus0004-varcall-with-mg-with_node.pdf', sep = "");
pdf(file)
plot(phylo, use.edge.length = FALSE,edge.color = edgecol, tip.color = tipcol, cex = 0.15)
nodelabels(cex=0.25, frame = "none", col = "blue")
dev.off()

# only showing relevant node labels
ldv_nodes = rep("", length(phylo$node.label))
ldv_nodes[as.numeric(names(which(unlist(lapply(ldv_i, length)) > 0)))] = 1:sum(unlist(lapply(ldv_i, length)) > 0)
file = paste("../figures/", format(Sys.time(), "%Y-%m-%d"), '_5261-Aus0004-varcall-with-mg-node_number.pdf', sep = "");
pdf(file)
plot(phylo, use.edge.length = T,edge.color = edgecol, tip.color = tipcol, cex = 0.25, show.tip.label = F)

nodelabels(text = ldv_nodes, cex = .25, frame = "circle", col = "blue")
dev.off()

# barplot showing proportion of alternative alleles detected
par(mar = c(5, 5, 5, 5))
bp = barplot(c(i_ldv_100, i_ldv_lt100, non_i_ldv_100, non_i_ldv_lt100), col = "black", ylab = "Proportion of alternative alleles\nmapped to Aus ref genome")
text(x = bp, y = -.05,
     labels = c(
       paste0("LDV pos: 100% freq\n(", nrow(output_df[output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency %in% 1,]), " / ",length(ldv_i_pos), ")"), 
       paste0("LDV pos: 100% freq\n(", nrow(output_df[output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency < 1 &output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency  > 0 ,]), " / ",length(ldv_i_pos), ")"), 
       paste0("non-LDV pos: 100% freq\n(", nrow(output_df[!output_df$LDV.Position %in% ldv_i_pos & output_df$ALT.Allele.Frequency %in% 1,]), " / ",length(unlist(vrefm_st_ldv[!unlist(vrefm_st_ldv) %in% ldv_i_pos])), ")"), 
       "grey"),
     xpd = T, 
     srt = 45)
```

